rasterio/__init__.py

```diff
@@ -190,46 +189,54 @@ def open(fp, mode='r', driver=None, width=None, height=None, count=None,
             "Blacklisted: file cannot be opened by "
             "driver '{0}' in '{1}' mode".format(driver, mode))
 
-    # Special case for file object argument.
+    # If the fp argument is a file-like object and can be adapted by
+    # rasterio's FilePath we do so. Otherwise, we use a MemoryFile to
+    # hold fp's contents and store that in an ExitStack attached to the
+    # dataset object that we will return. When a dataset's close method
+    # is called, this ExitStack will be unwound and the MemoryFile's
+    # storage will be cleaned up.
     if mode == 'r' and hasattr(fp, 'read'):
-
-        @contextmanager
-        def fp_reader(fp):
-            if have_vsi_plugin:
-                vsi_file = FilePath(fp)
-            else:
-                vsi_file = MemoryFile(fp.read())
-            dataset = vsi_file.open(driver=driver, sharing=sharing)
-            try:
-                yield dataset
-            finally:
-                dataset.close()
-                vsi_file.close()
-
-        return fp_reader(fp)
+        if have_vsi_plugin:
+            return FilePath(fp).open(driver=driver, sharing=sharing, **kwargs)
```

Comment:
Do we need to be concerned with the FilePath object being garbage collected at the GDAL Plugin being unregistered?

Reply:
I think see what you mean. I'll make a new issue about moving plugin registration to the _env module, around https://github.com/rasterio/rasterio/blob/master/rasterio/_env.pyx#L359.

Reply:
Here's the copying https://github.com/OSGeo/gdal/blob/35c07b18316b4b6d238f6d60b82c31e25662ad27/port/cpl_vsil_plugin.cpp#L111. I think we're safe.